pkg/health/checker/checker.go

```diff
@@ -41,28 +69,37 @@ type healthStore interface {
 	GetServiceHealth(service string) (map[string]consul.WatchResult, error)
 }
 
-type healthKV interface {
-	List(prefix string, opts *api.QueryOptions) (api.KVPairs, *api.QueryMeta, error)
+type ReplicationControllerStore interface {
+	Get(id rcfields.ID) (rcfields.RC, error)
 }
 
-type consulHealthChecker struct {
-	client      consulutil.ConsulClient
-	kv          healthKV
-	consulStore healthStore
+type LabelReader interface {
+	GetMatches(klabels.Selector, labels.Type) ([]labels.Labeled, error)
 }
-type consulHealth interface {
-	Node(types.NodeName, *api.QueryOptions) ([]*api.HealthCheck, *api.QueryMeta, error)
+
+type healthChecker struct {
+	healthClient   HealthClient
+	resourceClient ResourceClient
+	consulClient   consulutil.ConsulClient
+	kv             healthKV
+	consulStore    healthStore
+	rcStore        ReplicationControllerStore
+	labelReader    LabelReader
 }
 
-func NewConsulHealthChecker(client consulutil.ConsulClient) ConsulHealthChecker {
-	return consulHealthChecker{
-		client:      client,
-		kv:          client.KV(),
-		consulStore: consul.NewConsulStore(client),
+func NewHealthChecker(hClient HealthClient, resourceClient ResourceClient, cClient consulutil.ConsulClient, rcStore ReplicationControllerStore, labelReader LabelReader) HealthChecker {
```

Comment:
WDYT about a new constructor for clients who want to specify all of these arguments? The behaviour is special, so a special constructor isn't unreasonable. How about NewShadowTrafficHealthChecker(…). It has the benefit of being a bit clearer about intent and doesn't require the caller to specify so many `nil`s. At the very least, this constructor should declare which arguments are optional and which aren't.

Reply:
Are you suggesting to rename NewHealthChecker to NewShadowTrafficHealthChecker with optional parameters or to create an additional constructor.

Would it be implicitly or explicitly specified that certain parameters are optional? Golang doesn't seem to have a way to specify optional parameters unless the function takes a struct and the fields in the struct are optional I think.

Reply:
I think he's suggesting a totally separate constructor, like `NewShadowTrafficHealthChecker` and leave the old `NewHealthChecker` alone for now. Then when we're confident in the health API service we can do a rename and drop the old consul related arguments

Reply:
^ yeah, this sounds right. It's easier to find callers of the old constructor than it is to find callers of a shared constructor that use deprecated arguments.

Reply:
In all the call sites where there are several nils passed in to the constructor, I suggest using a constructor that doesn't require the caller to pass in those nils. The NewShadowTrafficHealthChecker can call into the main constructor and pass in the nil arguments itself. This is a style request, it doesn't have much bearing on correctness.

Here's an example: https://play.golang.org/p/SaCDXIJO13-

Reply:
Cool, thanks. Why the name "ShadowTraffic"?

Reply:
"shadowing" traffic is the technique of duplicating all of the traffic going to destination A and sending it also to destination B. Usually this is done to see if destination B can handle the load that destination A can, and it's done when you are evaluating replacing A with B.

That's basically what we're doing here, we want to see if the health service can handle being accessed every time we used to be looking in consul